lib/phoenix/socket/transport.ex

```diff
@@ -13,6 +13,11 @@ defmodule Phoenix.Socket.Transport do
   many conveniences that invokes the underlying socket API to make
   it easier to build custom transports.
 
+  Of note, this behaviour is a superset of the `WebSock` behaviour, and
+  as such all implementations of `Phoenix.Socket.Transport` are also
+  implementations of `WebSock`. It is not possible to note this in code
+  without raising warnings about conflicting implementations.
```

Comment:
Ah... I see now. Phoenix.Socket.Transport in Phoenix is generic. WebSock is an implementation of Phoenix.Socket.Transport for WebSockets + a generic abstraction for adapters. So now that I think about it, I don't think we should list `WebSock` here.

Reply:
I.e. we should just remove this whole paragraph and instead say that WebSock implements Phoenix.Socket.Transport.

Reply:
> I.e. we should just remove this whole paragraph and instead say that WebSock implements Phoenix.Socket.Transport.

Do you mean that we should say that here, just in a differently worded comment, or that `WebSock` should actually `@impl` Phoenix.Socket.Transport over in the WebSock project?

Reply:
Right, the thing is that Phoenix.Socket.Transport is meant to be transport agnostic, so it doesn't make sense to say it implements WebSock.

At the same time, we don't want WebSock to depend on Phoenix, because people can use it without Phoenix.

So I think we should just informally says in WebSocket that it implements Phoenix.Socket.Transport and that's it. WDYT?

Reply:
This is also the reason why I think it should exist in the Phoenix org (I was originally considering it should be Plug but I think it is more aligned to Phoenix).

Reply:
Happy to note it at both ends (in fact, I'm pretty sure I already do note the similarity in the WebSock docs, at least as a source of inspiration). My main worry (speaking from experience as I pieced together an understanding of all of Phoenix's internals) is that without some sort of note here, it's not at all obvious how a fundamental part of Phoenix's WebSocket implementation works. Maybe the note could instead live in Phoenix.Transport.WebSocket, as a code comment near the `WebSock.upgrade/4` call?

Reply:
I actually think that's better, TBH. As I've mentioned before, someone trying to understand how WebSockets work in Phoenix is going to check out Phoenix.Transports.WebSocket first (just given the name), and so as many breadcrumbs as possible should be left there

Reply:
As an aside, the ability for behaviours to compose other behaviours is something I've wished for on multiple occasions (a concrete example is ThousandIsland.Handler wanting to be a superset of GenServer). Has this been proposed before / are there plans for it / is it something you'd entertain in the language?

Reply:
> Maybe the note could instead live in Phoenix.Transport.WebSocket, as a code comment near the WebSock.upgrade/4 call?

Agreed!

> Has this been proposed before / are there plans for it / is it something you'd entertain a PR for in the language?

Nothing we can do. Behaviours are somewhat tied to Dialyzer and Dialyzer will complain anyway.